### PR TITLE
Fix final gc lowering on dynamically sized allocation

### LINKED ISSUE #553: lowerGCAllocBytes LLVM error running tests
@vchuravy told me to file this issue: seems just running the Enzyme tests with julia master is hitting an assertion error:
```
julia: /data/vtjnash/julia/usr/include/llvm/Support/Casting.h:269: typename llvm::cast_retty<X, Y*>::ret_type llvm::cast(Y*) [with X = llvm::ConstantInt; Y = llvm::Value
; typename llvm::cast_retty<X, Y*>::ret_type = llvm::ConstantInt*]: Assertion `isa<X>(Val) && "cast<Ty>() argument of incompatible type!"' failed.                       
                                                                                                                                                                         
[587423] signal (6.-6): Aborted                                                                                                                                          
in expression starting at /home/vtjnash/.julia/packages/Enzyme/7ekWs/test/abi.jl:4                                                                                       
gsignal at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)                                                                                                                
abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)                                                                                                                  
unknown function (ip: 0x7f08677bd728)                                               
__assert_fail at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)                                                                                                          
cast<llvm::ConstantInt, llvm::Value> at /data/vtjnash/julia/usr/include/llvm/Support/Casting.h:269 [inlined]                                                             
lowerGCAllocBytes at /data/vtjnash/julia/src/llvm-final-gc-lowering.cpp:208                                                                                                                                                                        ```

### PR DESCRIPTION
**Pull request by wsmoses**
Resolves https://github.com/EnzymeAD/Enzyme.jl/issues/553 and is hit during 1.9 on https://github.com/EnzymeAD/Enzyme.jl/issues/597 aka https://github.com/JuliaLang/julia/issues/48621

cc @vchuravy @vtjnash

### CODE REVIEW

**Comment by wsmoses** [comment]
This should be backported to 1.9 and earlier versions as well.

File: test/llvmpasses/final-lower-gc.ll [hunk 1]
```diff
@@ -67,6 +67,21 @@ top:
...
 }
 
+define {} addrspace(10)* @gc_alloc_lowering_var(i64 %size) {
+top:
+; CHECK-LABEL: @gc_alloc_lowering_var
+  %pgcstack = call {}*** @julia.get_pgcstack()
+  %ptls = call {}*** @julia.ptls_states()
+  %ptls_i8 = bitcast {}*** %ptls to i8*
+; CHECK: %0 = add i64 %size, 8
+; CHECK: %v = call noalias nonnull dereferenceable(8) {} addrspace(10)* @ijl_gc_big_alloc(i8* %ptls_i8, i64 %0)
```
Comment (vchuravy) [inline test/llvmpasses/final-lower-gc.ll:77]: IIUC. We should not be calling `jl_gc_big_alloc`, but rather `jl_gc_allocobj`.

Comment (vchuravy) [reply test/llvmpasses/final-lower-gc.ll:77]: Or rather `jl_gc_alloc_typed(ptls, size, NULL)`. To be precise we should re-use the ptls we obtained here, but `jl_gc_alloc` is not exported.

**Review by vchuravy — APPROVED** [review]

**Comment by vchuravy** [comment]
Also can you try locally `jl_gc_alloc` https://github.com/JuliaLang/julia/blob/05b99af2b78b0ee88e1e501d70e31430043dbbb5/src/gc.c#L3181 directly? It is exported from `libjulia-internal`.

**Comment by wsmoses** [comment]
The latter fails.

```julia
./julia/julia --project ni.jl 
rr: Saving execution to trace directory `/home/wmoses/.local/share/rr/julia-63'.

[1867731] signal (11.1): Segmentation fault
in expression starting at /home/wmoses/git/Enzyme.jl/ni.jl:20
jl_gc_pool_alloc_inner at /home/wmoses/git/Enzyme.jl/julia/src/gc.c:1295
jl_gc_pool_alloc_noinline at /home/wmoses/git/Enzyme.jl/julia/src/gc.c:1359
jl_gc_alloc_ at /home/wmoses/git/Enzyme.jl/julia/src/julia_internal.h:460
jl_gc_alloc at /home/wmoses/git/Enzyme.jl/julia/src/gc.c:3183
ijl_gc_alloc_typed at /home/wmoses/git/Enzyme.jl/julia/src/gc.c:3783
unknown function (ip: 0x7ff067e4c95c)
Allocations: 61395440 (Pool: 0; Other: 61395440); GC: 70
Allocations: 61395440 (Pool: 0; Other: 61395440); GC: 70
Segmentation fault
```

**Comment by wsmoses** [comment]
`jl_gc_alloc` succeeds, and now pushed code to use that.

**Comment by KristofferC** [comment]
I backported this but got an error in the llvmpasses: https://buildkite.com/julialang/julia-release-1-dot-9/builds/101#01866edd-e69e-4681-81ce-5585968bd21b. I'll drop it from the backport branch for now and either @wsmoses and @vchuravy can push a correct backport to those branches.

**Pull request merged**